windows/src/engine/keyman32/calldll.cpp

```diff
@@ -438,25 +438,61 @@ extern "C" BOOL _declspec(dllexport) WINAPI KMDisplayIM(HWND hwnd, BOOL FShowAlw
 	*Globals::hwndIM() = hwnd;
 	*Globals::hwndIMAlways() = FShowAlways;
 
-	POINT pt;
-	RECT r;
+	POINT IMTop, Caret;
+  RECT RectIM, RectApp;
+  HWND hFocus;
+  int n = 0;
+
+  HMONITOR hMonitor = MonitorFromWindow(hwnd, MONITOR_DEFAULTTONEAREST);
+	if (hMonitor == NULL) {
+		return FALSE;
+	}
+
+	MONITORINFO monitor_info;
+	monitor_info.cbSize = sizeof(monitor_info);
+	if (!GetMonitorInfo(hMonitor, &monitor_info)) {
+		return FALSE;
+	}
 
-	GetCaretPos(&pt);
+  if (!GetWindowRect(hwnd, &RectIM)) {
+    return FALSE;
+  }
 
-	int n;
-	//if(pt.x == 0 && pt.y == 0)
-		n = SWP_NOMOVE;
-	//else n = 0;
+  // Only adjust the IM window coordinates if it is not going to be visible on the
+  // monitor it is currently showing on. Otherwise leave it to the 3rd paryt app
+  // to control the location of the IM Window.
+  // If changing the coordinates place the IM window use the caret postion of the
+  // application that has focus.
+  if ((RectIM.left < monitor_info.rcMonitor.left) || (RectIM.right > monitor_info.rcMonitor.right) ||
+      (RectIM.top < monitor_info.rcMonitor.top) || (RectIM.bottom > monitor_info.rcMonitor.bottom)) {
 
-	ClientToScreen(hwndFocus, &pt);
-	GetWindowRect(hwnd, &r);
-	if(pt.x + r.right - r.left >= GetSystemMetrics(SM_CXSCREEN))
-		pt.x = GetSystemMetrics(SM_CXSCREEN) - (r.right - r.left);
+    if (!GetCaretPos(&Caret)) {
+      return FALSE;
+    }
 
-	if(pt.y - (r.bottom-r.top) > 0) pt.y -= (r.bottom-r.top);
-	else pt.y += 32; // guessing...
+    hFocus = GetFocus();
+    if (!GetWindowRect(hFocus, &RectApp)) {
+      return FALSE;
+    }
+
+    IMTop.x = Caret.x + RectApp.left;
+    IMTop.y = Caret.y + RectApp.top + 24;
+
+    if (IMTop.x + RectIM.right - RectIM.left > RectApp.right) {
```

Comment:
Theoretically the IM could be bigger than the app window, so we would end up that it starts to the left of the app window. If the app window is at 0 then the IM would still be partially invisible.

Reply:
Good catch I think it we can lock the IM to the app right edge but then if it is bigger we start on the app left edge and then let it extend pass the right edge of the app and possibly visible screen.

It is worth noting in the call back dll managing the imx extenisions it can also position the IM window. Which is why this code only kicks in if it is not visible at all.